### PR TITLE
[162] Build UI - Eval Form Step 2 - Evaluation Criteria Add/Remove Criteria

### LINKED ISSUE #162: Build UI - Eval Form Step 2 - Evaluation Criteria Add/Remove Criteria
# What/Why
This task allows the user to create the basics of the criteria as well as add and remove criteria on the form. After this task is complete, valid forms can be created and published, albeit using only the numeric scoring option.

## Acceptance Criteria
- [x] Initial/empty state:
Screenshot 2024-09-13 at 12 44 37 PM
- [x] Add Another Criteria Button adds a new, empty criteria to the form
- [x] Delete button removes the criteria from the form
  - [x] when the last criteria is remove the form will have one criteria in the empty/default state
- [x] Close button collapses the options to a single row view ~showing the Title~
  - [x] Expand displays the options
  Screenshot 2024-09-13 at 12 33 10 PM
- [x] `Title`, `Description`, `Points`, `Scoring Type` are required
- [x] Numeric type of scoring is selected and functional
  - Rating and Binary scale are out of scope, if selected nothing happens (the form is always :numeric scoring at this point)
- [x] Criteria Points / Weight are positive integers
  - no decimal or negative numbers may be entered

_Disregard the following if this work **does not** include any UI changes_
- [x] WAVE screenshot
- [ ] Automated a11y tests (axe-rspec specs)
- [x] Keyboard accessibility (tab navigation, form interactions)
- [x] Responsiveness at desktop and mobile

### LINKED ISSUE #164: Build UI - Eval Form Step 2 - Evaluation Criteria Binary Scale
# What/Why
This task adds the ability to add binary scale options to the form.

### Figma/Mockup
Screenshot 2024-09-13 at 1 17 10 PM

## Acceptance Criteria
- [x] When the user select **Binary scale**, the option is set on the criteria
- [x] Labels are required, and initially empty
  * empty label fails validation 

_Disregard the following if this work **does not** include any UI changes_
- [x] WAVE screenshot
- [ ] Automated a11y tests (axe-rspec specs)
- [x] Keyboard accessibility (tab navigation, form interactions)
- [x] Responsiveness at desktop and mobile

## Out of Scope
Tooltips #168

### LINKED ISSUE #165: Build UI - Eval Form Step 2 - Evaluation Criteria Rating Scale
# What/Why
This task adds the ability to add rating scale options to the form.

### Figma/Mockup
Screenshot 2024-09-13 at 1 24 08 PM

## Acceptance Criteria
- [x] matches the mockup

_Disregard the following if this work **does not** include any UI changes_
- [x] WAVE screenshot
- [ ] Automated a11y tests (axe-rspec specs)
- [x] Keyboard accessibility (tab navigation, form interactions)
- [x] Responsiveness at desktop and mobile

## Out of Scope
Tooltips #168

### PR DESCRIPTION
**Pull request by cpreisinger**
#61 
#162 
#164 
#165
#167

### CODE REVIEW

File: db/migrate/20241017172408_remove_unique_title_index_from_evaluation_criteria.rb
Comment (stepchud) [inline db/migrate/20241017172408_remove_unique_title_index_from_evaluation_criteria.rb:4]: is there a reason to remove this index?

File: app/javascript/evaluation_criteria.js
Comment (stepchud) [inline app/javascript/evaluation_criteria.js]: isn't this already declared on line 3?

File: db/migrate/20241017172408_remove_unique_title_index_from_evaluation_criteria.rb
Comment (stepchud) [reply db/migrate/20241017172408_remove_unique_title_index_from_evaluation_criteria.rb:4]: oh I remember discussing this, about deleting existing criteria and adding a new one with the same name 👌

**Review by stepchud — COMMENTED** [review]
there are a few specs failing

File: app/javascript/evaluation_criteria.js
Comment (stepchud) [inline app/javascript/evaluation_criteria.js:1]: I'm feeling a little uncomfortable with using vanilla JS on the criteria and stimulus-js (and USWDS-js) on the rest of the form. Stimulus really isn't a huge lift in terms of learning curve so I'd like to see if you can make any progress or at least identify the specific challenges that make it difficult to use on these form elements. I won't ultimately block on this requirement but I want to spend some more time on it. It's worth learning now so it will be easier to build and maintain later.

Comment (stepchud) [reply app/javascript/evaluation_criteria.js:1]: Checkout the handbook and documentation: https://stimulus.hotwired.dev/handbook/hello-stimulus

File: .circleci/config.yml
Comment (stepchud) [inline .circleci/config.yml]:
```suggestion
```

File: app/assets/stylesheets/application.sass.scss
Comment (stepchud) [inline app/assets/stylesheets/application.sass.scss]: remove comments before merge

File: app/javascript/controllers/evaluation_criteria_controller.js
Comment (stepchud) [inline app/javascript/controllers/evaluation_criteria_controller.js:23]: should this decrement `this.counter`?

File: app/javascript/controllers/evaluation_criteria_controller.js
Comment (stepchud) [inline app/javascript/controllers/evaluation_criteria_controller.js]: remove console.log

File: app/javascript/controllers/evaluation_criteria_controller.js
Comment (stepchud) [inline app/javascript/controllers/evaluation_criteria_controller.js]: this is fine, but for me it reads a little easier without negating the expression
```suggestion
        input.disabled = index < start || index > end;
```

File: app/javascript/controllers/evaluation_criteria_controller.js
Comment (stepchud) [inline app/javascript/controllers/evaluation_criteria_controller.js]: TIL `checkValidity()` and `reportValidity()`!

File: app/views/evaluation_forms/_evaluation_criterion_fields.html.erb
Comment (stepchud) [inline app/views/evaluation_forms/_evaluation_criterion_fields.html.erb:67]: it looks like some of these data -target attributes are not being used, like this one?

File: db/structure.sql
Comment (stepchud) [inline db/structure.sql]: I just recreated my DB from phoenix and I think this field type is now text[] (meaning you can revert and not commit this change when it shows up in your diff), although I'm not sure if it makes much difference for the apps. It will just be churning back and forth depending on each devs local schema. Not a blocker for now, but we will eventually want to get the local dev schemas in sync.

File: spec/factories/evaluation_criteria.rb
Comment (stepchud) [inline spec/factories/evaluation_criteria.rb:35]: 👍 nice

File: db/structure.sql
Comment (stepchud) [reply db/structure.sql]: I went ahead and reverted this change on the branch

**Review by stepchud — COMMENTED** [review]
This can be merged, just left a few cleanup comments. The functionality is working afaict. Code structure is clean and readable. 🙌

File: app/javascript/controllers/evaluation_criteria_controller.js
Comment (cpreisinger) [reply app/javascript/controllers/evaluation_criteria_controller.js:23]: It doesn't really matter since it's mostly there for making sure the submitted nested fields have a unique ID for each criteria. It could cause conflicts with that uniqueness when decrementing. 

Removed fields that were presenting when updating remain in the form with a hidden `id` and `_destroy` field to make sure they get removed from the database.

File: app/javascript/controllers/evaluation_criteria_controller.js
Comment (cpreisinger) [reply app/javascript/controllers/evaluation_criteria_controller.js]: I think `reportValidity()` does both conditional and the html5 popup I can probably simplify this

File: app/views/evaluation_forms/_evaluation_criterion_fields.html.erb
Comment (cpreisinger) [reply app/views/evaluation_forms/_evaluation_criterion_fields.html.erb:67]: Yeah I'll go through and clean these. This was when I was initially thinking of approaching the Stimulus thing a different way. Which might still work but I didn't try it because I was skeptical. The multiple of the same controller scoped to its own targets approach for the criteria and then one controller for the list itself.

File: app/javascript/controllers/evaluation_criteria_controller.js
Comment (stepchud) [reply app/javascript/controllers/evaluation_criteria_controller.js:23]: gotcha, that works 👍

File: app/views/evaluation_forms/confirmation.html.erb
Comment (stepchud) [inline app/views/evaluation_forms/confirmation.html.erb]:
```suggestion
      During the evaluation period, the form will be available to your evaluators, 
```

**Review by stepchud — APPROVED** [review]
looks good to me

**Pull request merged**
